src/client/clientenvironment.cpp

```diff
@@ -262,29 +259,31 @@ void ClientEnvironment::step(float dtime)
 		Step active objects and update lighting of them
 	*/
 
-	g_profiler->avg("CEnv: num of objects", m_active_objects.size());
 	bool update_lighting = m_active_object_light_update_interval.step(dtime, 0.21);
-	for (auto &ao_it : m_active_objects) {
-		ClientActiveObject* obj = ao_it.second;
+	auto cb_state = [this, dtime, update_lighting, day_night_ratio] (ClientActiveObject *cao) {
 		// Step object
-		obj->step(dtime, this);
+		cao->step(dtime, this);
 
 		if (update_lighting) {
 			// Update lighting
 			u8 light = 0;
 			bool pos_ok;
 
 			// Get node at head
-			v3s16 p = obj->getLightPosition();
-			MapNode n = m_map->getNodeNoEx(p, &pos_ok);
+			v3s16 p = cao->getLightPosition();
+			MapNode n = this->m_map->getNodeNoEx(p, &pos_ok);
 			if (pos_ok)
 				light = n.getLightBlend(day_night_ratio, m_client->ndef());
 			else
 				light = blend_light(day_night_ratio, LIGHT_SUN, 0);
 
-			obj->updateLight(light);
+			cao->updateLight(light);
 		}
-	}
+	};
+
+	// hold the callback state
+	std::function<void(ClientActiveObject *)> cb = std::ref(cb_state);
+	m_ao_manager.step(dtime, cb);
```

Comment:
You should be able to pass directly cb_state (same signature than cb, `void(ClientActiveObject *)`)

Reply:
i cannot it's why i added std::ref, compiler refused it due to lambda capture. Without capture it's allowed

Reply:
If you can't, then remove `auto` and just do `std::function<void(ClientActiveObject *)> cb_state = ...`